### DISCUSSION TITLE
Discussion: string interpolation with char span

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#3375] [Language Ideas]
The runtime already provides `Concat` span overloads up to four arguments,
```cs
$"{span[..4]}***{span[^4..]}"
```
```cs
string.Concat(span[..4], "***", span[^4..])
```
For language support we probably need some other mechanism to support >4 interpolations.

**Comment by canton7** [comment]
Don't forget that string interpolation lowers to `String.Concat` in only a few cases: most of the time it lowers to `String.Format`. Don't forget `FormattableString` too. 

If you have a rule that `Span<char>` gets turned into a string (and that's used for `FormattableString` too), then maybe that's the way forwards.

**Comment by alrz** [comment]
>If you have a rule that Span<char> gets turned into a string

That's ToString. The point is to not create a string just for concat. Note stirng.Concat directly accepts spans.

I suppose no format specifier would be allowed for such string much like const string interpolations.

**Comment by canton7** [comment]
Again, string interpolation is not sugar for concatenation. It's sugar for `string.Format`. It only gets lowered to concatenation in very specific circumstances. If you want to allow spans in interpolated strings, they need to work in *all* cases, including when `String.Format` is used and when `FormattableString` is used.

At face value this is obviously not possible. `FormattableString`, lets you fetch the arguments as an object array. `String.Format` lets you use an `ICustomFormatter`. That's a complete no-go.

Hence my suggestion: the only way I could think through this is to define string interpolation as converting a `Span<char>` to a string, and then everything else happens on that string. `FormattableString` sees a string. `ICustomFormatter` sees a string. You don't have a performance gain but you have gained convenience.

Then you apply the optimization where, if the string interpolation can be lowered to concatenation, and other conditions are met, you elide the step of turning the span into a string, and you use the `String.Concat` overload which takes spans.

That also gets around your problem of what to do you've got more than 4 things to concat: you fall back to turning each span into a string, and you lose the optimization.
